// @bangbang93/utils/nestjs/base-crud-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseCrudService = void 0;
const index_1 = require("../index");
const mongodb_1 = require("../mongodb");
class BaseCrudService {
    constructor(model) {
        this.model = model;
    }
    async getById(id) {
        return this.model.findById(id);
    }
    async listByIds(ids) {
        return this.model.find({ _id: { $in: ids.map(mongodb_1.toObjectId) } });
    }
    async create(data) {
        return this.model.create(data);
    }
    async search(query, skip = 0, limit = index_1.DEFAULT_LIMIT, queryHelper) {
        return (0, mongodb_1.findAndCount)(this.model, query, skip, limit, queryHelper);
    }
    async update(id, data) {
        await this.model.updateOne({ _id: (0, mongodb_1.toObjectId)(id) }, {
            $set: data,
        });
    }
    async delete(id) {
        await this.model.deleteOne({ _id: (0, mongodb_1.toObjectId)(id) });
    }
}
exports.BaseCrudService = BaseCrudService;
