// @bangbang93/utils/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.trimDeep = exports.toBoolean = exports.Deprecated = exports.DEFAULT_LIMIT = exports.DEFAULT_PAGE = exports.DEFAULT_SKIP = exports.second = void 0;
const is_1 = __importDefault(require("@sindresorhus/is"));
const lodash_1 = require("lodash");
const ms_1 = __importDefault(require("ms"));
const util_1 = require("util");
function second(str) {
    return ~~((0, ms_1.default)(str) / 1000);
}
exports.second = second;
exports.DEFAULT_SKIP = 0;
exports.DEFAULT_PAGE = 1;
exports.DEFAULT_LIMIT = 10;
function Deprecated(message) {
    return (target, propertyKey, descriptor) => {
        if (descriptor.get) {
            descriptor.get = (0, util_1.deprecate)(descriptor.get, message);
        }
        else if (typeof descriptor.value === 'function') {
            descriptor.value = (0, util_1.deprecate)(descriptor.value, message);
        }
    };
}
exports.Deprecated = Deprecated;
function toBoolean(value) {
    if ((0, lodash_1.isNil)(value))
        return value;
    return [true, 'true', '1', 'yes', 1].includes(value);
}
exports.toBoolean = toBoolean;
function trimDeep(obj) {
    return (0, lodash_1.mapValues)(obj, (v) => {
        if (is_1.default.string(v)) {
            return v.trim();
        }
        if (is_1.default.plainObject(v)) {
            return trimDeep(v);
        }
        return v;
    });
}
exports.trimDeep = trimDeep;
__exportStar(require("./types"), exports);
