// @bangbang93/utils/nestjs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiError = exports.InjectRef = exports.IntParam = exports.ApiFile = exports.PagedResDto = exports.IdDto = exports.SortablePagedDto = exports.PagedDto = exports.getDefaultPageLimit = exports.setDefaultPageLimit = exports.generateSwagger = exports.ApiSummary = void 0;
const common_1 = require("@nestjs/common");
const core_1 = require("@nestjs/core");
const swagger_1 = require("@nestjs/swagger");
const class_validator_1 = require("class-validator");
const fs_1 = require("fs");
const path = __importStar(require("path"));
const ApiSummary = (summary) => (0, swagger_1.ApiOperation)({ summary });
exports.ApiSummary = ApiSummary;
async function generateSwagger(appModule, options = {}) {
    const app = await core_1.NestFactory.create(appModule);
    const builder = new swagger_1.DocumentBuilder()
        .setTitle(options?.title ?? 'app')
        .setDescription(options?.description ?? '')
        .setVersion(options?.version ?? '');
    if (options.prefix) {
        app.setGlobalPrefix(options.prefix);
    }
    if (options.servers) {
        for (const server of options.servers) {
            builder.addServer(server.url, server.description, server.variables);
        }
    }
    const document = swagger_1.SwaggerModule.createDocument(app, builder.build());
    const outputPath = options.outputPath ?? path.resolve(process.cwd(), 'swagger.json');
    (0, fs_1.writeFileSync)(outputPath, JSON.stringify(document), { encoding: 'utf8' });
    await app.close();
}
exports.generateSwagger = generateSwagger;
let DefaultPageLimit = 10;
const MAX_PAGE = 500;
function setDefaultPageLimit(limit) {
    DefaultPageLimit = limit;
}
exports.setDefaultPageLimit = setDefaultPageLimit;
function getDefaultPageLimit() {
    return DefaultPageLimit;
}
exports.getDefaultPageLimit = getDefaultPageLimit;
class PagedDto {
    constructor() {
        this.page = 1;
        this.limit = getDefaultPageLimit();
    }
    get skip() {
        return (this.page - 1) * this.limit;
    }
    set skip(value) {
        // noop
    }
}
__decorate([
    (0, swagger_1.ApiPropertyOptional)(),
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsInt)(),
    (0, class_validator_1.Min)(1),
    __metadata("design:type", Number)
], PagedDto.prototype, "page", void 0);
__decorate([
    (0, swagger_1.ApiPropertyOptional)(),
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.IsInt)(),
    (0, class_validator_1.Min)(1),
    (0, class_validator_1.Max)(MAX_PAGE),
    __metadata("design:type", Number)
], PagedDto.prototype, "limit", void 0);
exports.PagedDto = PagedDto;
class SortablePagedDto extends PagedDto {
}
__decorate([
    (0, swagger_1.ApiPropertyOptional)({ description: '例： +createdAt,"+-"代表正序倒序' }),
    (0, class_validator_1.IsOptional)(),
    (0, class_validator_1.Matches)(/^[+-].+$/),
    __metadata("design:type", String)
], SortablePagedDto.prototype, "sort", void 0);
exports.SortablePagedDto = SortablePagedDto;
class IdDto {
}
__decorate([
    (0, swagger_1.ApiProperty)(),
    (0, class_validator_1.IsMongoId)(),
    __metadata("design:type", String)
], IdDto.prototype, "id", void 0);
exports.IdDto = IdDto;
function PagedResDto(constructor) {
    const name = `Paged${constructor.name}`;
    class PagedRes {
    }
    __decorate([
        (0, swagger_1.ApiProperty)(),
        __metadata("design:type", Number)
    ], PagedRes.prototype, "count", void 0);
    __decorate([
        (0, swagger_1.ApiProperty)({ type: [constructor] }),
        __metadata("design:type", Array)
    ], PagedRes.prototype, "data", void 0);
    Reflect.defineProperty(PagedRes, 'name', {
        writable: false,
        enumerable: false,
        configurable: true,
        value: name,
    });
    return PagedRes;
}
exports.PagedResDto = PagedResDto;
function ApiFile(fileName = 'file') {
    return (0, common_1.applyDecorators)((0, swagger_1.ApiConsumes)('multipart/form-data'), (0, swagger_1.ApiBody)({
        schema: {
            type: 'object',
            properties: {
                [fileName]: {
                    type: 'file',
                    format: 'blob',
                },
            },
        },
    }));
}
exports.ApiFile = ApiFile;
function IntParam(name) {
    return (target, propertyKey, parameterIndex) => {
        (0, common_1.Param)(name, common_1.ParseIntPipe)(target, propertyKey, parameterIndex);
    };
}
exports.IntParam = IntParam;
function InjectRef(fn) {
    return (0, common_1.Inject)((0, common_1.forwardRef)(fn));
}
exports.InjectRef = InjectRef;
/**
 * Api错误响应装饰器
 * @param errors 错误定义
 * @param code 错误码
 * @param message 错误信息
 * @param description 描述
 * @example @ApiError(ServiceErrors, 'COMMON_NO_SUCH_OBJECT', '找不到对象')
 */
function ApiError(errors, code, message, description = message) {
    const error = errors[code];
    return (0, common_1.applyDecorators)((0, swagger_1.ApiResponse)({
        status: error[1],
        description,
        schema: {
            properties: {
                code: {
                    type: 'string',
                    example: code,
                },
                message: {
                    type: 'string',
                    example: message ?? error[0],
                },
            },
        },
    }));
}
exports.ApiError = ApiError;
